nibabel/cifti2/cifti2.py

```diff
@@ -0,0 +1,1095 @@
+        for p in self:
+            if isinstance(p, Cifti2Volume):
+                return p
+        return None
+
+    @volume.setter
+    def volume(self, volume):
+        if not isinstance(volume, Cifti2Volume):
+            raise ValueError("You can only set a volume with a volume")
+        for i, v in enumerate(self):
+            if isinstance(v, Cifti2Volume):
+                break
+        else:
+            self.append(volume)
+            return
+        self[i] = volume
+
+    @volume.deleter
+    def volume(self):
+        for i, v in enumerate(self):
+            if isinstance(v, Cifti2Volume):
+                break
+        else:
+            raise ValueError("No Cifti2Volume element")
+        del self[i]
+
+    @property
+    def brain_models(self):
+        for p in self:
+            if isinstance(p, Cifti2BrainModel):
+                yield p
+
+    def _to_xml_element(self):
+        if self.applies_to_matrix_dimension is None:
+            raise CIFTI2HeaderError(
+                'MatrixIndicesMap element requires to be applied to at least 1 dimension'
+            )
+
+        mat_ind_map = xml.Element('MatrixIndicesMap')
+        dims_as_strings = [str(dim) for dim in self.applies_to_matrix_dimension]
+        mat_ind_map.attrib['AppliesToMatrixDimension'] = ','.join(dims_as_strings)
+        for key in ['IndicesMapToDataType', 'NumberOfSeriesPoints', 'SeriesExponent',
+                    'SeriesStart', 'SeriesStep', 'SeriesUnit']:
+            attr = _underscore(key)
+            value = getattr(self, attr)
+            if value is not None:
+                mat_ind_map.attrib[key] = str(value)
+        for map_ in self:
+            mat_ind_map.append(map_._to_xml_element())
+
+        return mat_ind_map
+
+
+class Cifti2Matrix(xml.XmlSerializable, collections.MutableSequence):
+    def __init__(self):
+        self._mims = []
+        self.metadata = None
+
+    @property
+    def metadata(self):
+        return self._meta
+
+    @metadata.setter
+    def metadata(self, meta):
+        """ Set the metadata for this Cifti2Header
+
+        Parameters
+        ----------
+        meta : Cifti2MetaData
+
+        Returns
+        -------
+        None
+        """
+        self._meta = _value_if_klass(meta, Cifti2MetaData)
+
+    def __setitem__(self, key, value):
+        if not isinstance(value, Cifti2MatrixIndicesMap):
+            raise TypeError("Not a valid Cifti2MatrixIndicesMap instance")
+        self._mims[key] = value
+
+    def __getitem__(self, key):
+        return self._mims[key]
+
+    def __delitem__(self, key):
+        del self._mims[key]
+
+    def __len__(self):
+        return len(self._mims)
+
+    def insert(self, index, value):
+        if not isinstance(value, Cifti2MatrixIndicesMap):
+            raise TypeError("Not a valid Cifti2MatrixIndicesMap instance")
+        self._mims.insert(index, value)
+
+    def _to_xml_element(self):
+        if (len(self) == 0 and self.metadata is None):
```

Comment:
Thanks @coalsont I can add those without trouble to the matrix element. Just one question in the case of the three cases of cifti which are, for instance, conn by conn by several scalars, which nifti dimensions are used? Are you still discarding the first 4 ones?